src/Data/HasOrigin.php

```diff
@@ -10,6 +10,7 @@ trait HasOrigin
      * @var string
      */
     protected $origin;
+    protected $cachedHasOrigin = false;
```

Comment:
For properties that exist just for caching, I don't love initializing them with a value.

Can you remove the `false` here, and then refactor `hasOrigin()` so that it sets it to `true`/`false` appropriately?